Drone/drone_controller.py

```diff
@@ -0,0 +1,127 @@
+# Standard Library
+import coloredlogs
+import heapq
+import logging
+import threading
+from time import sleep
+import traceback
+import sys
+
+# Ours
+from drone import Drone
+from drone_controller_base import DroneControllerBase
+from ..Instructions.movement_instruction import MovementInstruction
+from ..Tasks.hover_task import HoverTask
+from ..Tasks.movement_task import MovementTask
+from ..Tasks.takeoff_task import TakeoffTask
+from ..Utilities import constants as c
+from ..Utilities import drone_exceptions as de
+from ..Utilities.lock import SharedLock
+
+class DroneController(DroneControllerBase):
+    """
+    Concrete implementation of DroneControllerBase. See drone_controller_base.py for
+    documentation.
+    """
+    def __init__(self, drone):
+        super(DroneController, self).__init__(drone)
+        self.logger = logging.getLogger(__name__)
+        self.debug = False
+        coloredlogs.install(level='DEBUG')
+        # The following two lines are purely for testing purposes. Instructions
+        # will be pushed onto the heap as a result of the swarm controller
+        # sending instructions or inter-drone communication.
+        heapq.heappush(self.instruction_queue, (0, MovementInstruction((5, 5, 0))))
+        heapq.heappush(self.instruction_queue, (0, MovementInstruction((-5, -5, 0))))
+
+    def set_id(self):
+        return 1
+
+    def update(self):
+        if self.task is not None:
+            # Do one iteration of whichever task we are in
+            result = self.task.do()
+
+            # If we are done, set the task to None so that 
+            # we can move on to the next instruction
+            if result:
+                self.task = None
+
+        if type(self.task) is HoverTask or self.task is None:
+            # Process remaining instructions
+            if len(self.instruction_queue):
+                # Stop hovering, if we were doing so
+                if type(self.task) is HoverTask:
+                    stop_hover_event = self.task.exit_task()
+                    stop_hover_event.wait()
+                    self.task = None
+                self.read_next_instruction()
+            # If there are no instructions, begin to hover
+            else:
+                if self.task is None:
+                    self.task = HoverTask(self.drone)
+         
+    def run_loop(self):
```

Comment:
This should be begin or something since it is basically the start, it should also be implemented such that it would be run using a `with` statement

Reply:
When the object is instanciated, is there any reason why `run_loop()` should not be run?

Reply:
It should always be called, otherwise object does nothing. I have not used `with` statements much so I'm not sure how it should be used in this case. Could you give an example of what you mean?

Reply:
basically a `with` statement will, when used, create an instance of the object, `DroneController` in this case, and then run some functions on it.
It basically encapsulates the creation and deletion of an object. You can choose whether or not you feel it is necessary.
However, for `DroneController`, the class may as well just run the function `run_loop` when the object is created (in `__init__`), since as you said, if the function is not called, the object does literally nothing.